Do not allow zerovalue as total_increasing for homewizard sensors

jbouwh opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I had some issues with my homewizard watermeter (local API fw 2.03) where the total water usage readout was reported as `0` where there was an error reading it out. In those cases a sensor could be unknown. If a `0` is processed and later the correct value is reported, the total sum be increased with the total usage. This need to be addressed. A way to address this as a work-a-round within Home assistant is to make the value `unknown`. That is what this PR tries to do.

An alternative is to make sure the API guarantees increasing values which is doesn't seem to do at currently, there is a big risk that this disturbs long term statistics for the sensors reporting wrong values. I had to correct my database to fix these.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #89021
- This PR is related to issue: #87432
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jbouwh committed:
Do not allow zerovalue  as total_increasing

jbouwh committed:
Fix for solar exports

DCSBL commented:
Thank you for this PR. I agree that we should fix this, but what is strange is that it should happen at all.

Anyway, I'm not sure if we should fix this in HA or in https://github.com/DCSBL/python-homewizard-energy. This seems to be an issue with the implementation of the API, not with HA. What do you think?

jbouwh commented:
> Thank you for this PR. I agree that we should fix this, but what is strange is that it should happen at all.
> 
> Anyway, I'm not sure if we should fix this in HA or in https://github.com/DCSBL/python-homewizard-energy. This seems to be an issue with the implementation of the API, not with HA. What do you think?

I think the guard should be in the API. It could just raise and HA can reraise an Update error. By not returning but raising the entities become unavailable. If later the connection restores the sensors become available again. Returning 0 is the worst you can do as HA will treat this as a reset and will take the next delta to be added to the sum.

jbouwh commented:
@DCSBL Is the zero value issue addressed in the mean time, or is this PR still needed?

frenck reviewed (approved):
It doesn't harm, lets merge it in to prevent issues and remove it when confirmed fixed.

../Frenck

jbouwh commented on homeassistant/components/homewizard/sensor.py:
```diff
@@ -386,7 +386,7 @@ async def async_setup_entry(
         HomeWizardSensorEntity(coordinator, entry, description)
         for description in SENSORS
-        if description.value_fn(coordinator.data.data) is not None
+        if getattr(coordinator.data.data, description.key, None) is not None
```
As `value_fn` is set for all instances, this does not need to change.
```suggestion
        if description.value_fn(coordinator.data.data) is not None
```

jbouwh committed:
Revert not needed change

frenck reviewed (approved):
Thanks, @jbouwh 👍 

../Frenck

DCSBL commented:
Sorry, forgotten about this one. Thanks 😄

frenck committed:
Adjust handling, making tests pass

frenck commented:
I've rebased the PR, and fixed the logic. Made it more explicit as well by adding it to the entity descriptions.

This is a bit more expressive, but it makes each entity description have its own logic.

Tests now pass, tested against actual devices as well.

@jbouwh Please take a look; and I hope you don't mind I jumped in. 🙏 

../Frenck

jbouwh commented:
> I've rebased the PR, and fixed the logic. Made it more explicit as well by adding it to the entity descriptions.
> 
> This is a bit more expressive, but it makes each entity description have its own logic.
> 
> Tests now pass, tested against actual devices as well.
> 
> @jbouwh Please take a look; and I hope you don't mind I jumped in. 🙏
> 
> ../Frenck

Not at all, nice work!

> I've rebased the PR, and fixed the logic. Made it more explicit as well by adding it to the entity descriptions.
> 
> This is a bit more expressive, but it makes each entity description have its own logic.
> 
> Tests now pass, tested against actual devices as well.
> 
> @jbouwh Please take a look; and I hope you don't mind I jumped in. 🙏
> 
> ../Frenck

Thnx, I'll have a look this morning. Nice the tests piss now.
Good work @frenck 👍

DCSBL reviewed (approved):
Simple and effective, thank you both!

jbouwh commented:
@frenck I think this can be merged now. We won't see increasing sensors reporting `0` anymore, thanx for your help, that was very appreciated!
